### PR TITLE
Add end to end tests

### LINKED ISSUE #29: Astarte action e2e tests
[Astarte action](https://github.com/astarte-platform/setup-astarte-action)
Create GitHub action using Astarte cluster workflow.

### PR DESCRIPTION
**Pull request by osmanhadzic**
Add end-to-end test for:

- Connect the device to Astarte
- Send property from device to server and from server to device
- Send Datastream from device to server and from server to device
- Send Aggregate data from device to server and from server to device
- Disconnect the device to Astarte

Closes #29

### CODE REVIEW

File: AstarteDeviceSDKCSharpE2E.Tests/Utilities/AstarteDeviceSingleton.cs [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+                    }
+
+                    string cryptoStoreDir = string.Empty;
+
+                    if (String.IsNullOrEmpty(cryptoStoreDir))
+                    {
+                        cryptoStoreDir = AppDomain.CurrentDomain.BaseDirectory;
+                    }
+
+
```
Comment (nedimtokic) [inline AstarteDeviceSDKCSharpE2E.Tests/Utilities/AstarteDeviceSingleton.cs:56]: Double new line.

File: AstarteDeviceSDKCSharpE2E.Tests/Utilities/AstarteDeviceSingleton.cs [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+                    AstarteDevice astarteDevice = new AstarteDevice(
+                        astarteMockData.DeviceId,
+                        astarteMockData.Realm,
+                        astarteMockData.CredentialsSecret,
+                        new MockInterfaceProvider(),
+                        astarteMockData.PairingUrl,
+                        cryptoStoreDir
+                    );
+
```
Comment (nedimtokic) [inline AstarteDeviceSDKCSharpE2E.Tests/Utilities/AstarteDeviceSingleton.cs:65]: Double new line.

File: AstarteDeviceSDKCSharpE2E.Tests/Utilities/AstarteMockDataTimestamp.cs [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+            public List<long> value { get; set; }
+        }
+
+        public class LongintegerEndpoint
+        {
+            public DateTime reception_timestamp { get; set; }
+            public DateTime timestamp { get; set; }
+            public long value { get; set; }
+        }
+
```
Comment (nedimtokic) [inline AstarteDeviceSDKCSharpE2E.Tests/Utilities/AstarteMockDataTimestamp.cs:154]: Double new line.

File: AstarteDeviceSDKCSharpE2E.Tests/AstarteDeviceTest.cs [hunk 4]
```diff
@@ -0,0 +1,405 @@
...
+            "/sensorUuid",
+            data);
+
+    }
+
+    /// <summary>
+    /// Test disconnect device from Astarte
+    /// </summary>
+    [Fact]
+    public void DisconnectDeviceFtomAstarte()
```
Comment (nedimtokic) [inline AstarteDeviceSDKCSharpE2E.Tests/AstarteDeviceTest.cs:324]: Fix the typo in `DisconnectDeviceFtomAstarte()`.

File: AstarteDeviceSDKCSharpE2E.Tests/AstarteMockData/AstarteMockDevice.cs [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+        (DoubleEndpoint, DoublearrayEndpoint,
+         IntegerEndpoint, IntegerarrayEndpoint,
+         BooleanEndpoint, BooleanarrayEndpoint,
+         LongintegerEndpoint, LongintegerarrayEndpoint,
+         StringEndpoint, StringarrayEndpoint,
+         DatetimeEndpoint, DatetimearrayEndpoint,
+         BinaryblobEndpoint, BinaryblobarrayEndpoint)
+        .GetHashCode();
+
+        private bool ComperByteArrays(byte[][] arraByte)
```
Comment (nedimtokic) [inline AstarteDeviceSDKCSharpE2E.Tests/AstarteMockData/AstarteMockDevice.cs:257]: Fix the typo in `ComperByteArrays` and `arraByte`.

**Review by nedimtokic — CHANGES_REQUESTED** [review]

File: AstarteDeviceSDKCSharpE2E.Tests/AstarteMockData/AstarteDeviceMockData.cs [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+using System;
+using System.IO;
+using System.Threading.Tasks;
+using AstarteDeviceSDKCSharp.Device;
+using AstarteDeviceSDKCSharpE2E.Tests.AstarteMockData;
+using AstarteDeviceSDKCSharpE2E.Tests.Utilities;
```
Comment (nedimtokic) [inline AstarteDeviceSDKCSharpE2E.Tests/AstarteMockData/AstarteDeviceMockData.cs:26]: Remove unused imports.

File: AstarteDeviceSDKCSharpE2E.Tests/AstarteMockData/AstarteMockDevice.cs [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+            DeviceId = Environment.GetEnvironmentVariable("E2E_DEVICE_ID",
+            EnvironmentVariableTarget.Process);
+            CredentialsSecret = Environment.GetEnvironmentVariable("E2E_CREDENTIALS_SECRET",
+            EnvironmentVariableTarget.Process);
+            ApiUrl = Environment.GetEnvironmentVariable("E2E_API_URL",
+            EnvironmentVariableTarget.Process);
+            AppEngineToken = Environment.GetEnvironmentVariable("E2E_TOKEN",
+            EnvironmentVariableTarget.Process);
+
+            InterfacesFld = Path.Combine
```
Comment (harlem88) [inline AstarteDeviceSDKCSharpE2E.Tests/AstarteMockData/AstarteMockDevice.cs:64]:
```suggestion
            InterfacesDir = Path.Combine
```

**Review by harlem88 — CHANGES_REQUESTED** [review]

File: AstarteDeviceSDKCSharpE2E.Tests/AstarteMockData/AstarteMockDevice.cs (hunk 7 above)
Comment (osmanhadzic) [reply AstarteDeviceSDKCSharpE2E.Tests/AstarteMockData/AstarteMockDevice.cs]: I made changes.

File: AstarteDeviceSDKCSharpE2E.Tests/AstarteAggregateDeviceTest.cs [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+
+        /// <summary>
+        /// Test for aggregated object datastreams in the direction from server to device.
+        /// </summary>
+        /// <returns></returns>
+        [Fact, TestPriority(5)]
+        public async void AggregateFromServerToDevice()
+        {
+            Console.WriteLine("Test for aggregated object datastreams"
+            + "in the direction from server to device.");
```
Comment (nedimtokic) [inline AstarteDeviceSDKCSharpE2E.Tests/AstarteAggregateDeviceTest.cs:112]:
```suggestion
            + " in the direction from server to device.");
```

File: AstarteDeviceSDKCSharpE2E.Tests/AstarteDatastreamDeviceTests.cs [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+        /// <summary>
+        /// Test for individual datastreams in the direction from device to server.
+        /// </summary>
+        /// <returns></returns>
+        [Fact, TestPriority(2)]
+        public async Task DatastreamFromDeviceToServer()
+        {
+            string interfaceName = astarteMockDevice.InterfaceDeviceData;
+            Console.WriteLine("Test for individual datastreams"
+            + "in the direction from device to server.");
```
Comment (nedimtokic) [inline AstarteDeviceSDKCSharpE2E.Tests/AstarteDatastreamDeviceTests.cs:71]:
```suggestion
            + " in the direction from device to server.");
```

File: AstarteDeviceSDKCSharpE2E.Tests/AstarteDatastreamDeviceTests.cs [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+        /// <summary>
+        /// Test for individual datastreams in the direction from server to device.
+        /// </summary>
+        /// <returns></returns>
+        [Fact, TestPriority(1)]
+        public async Task DatastreamFromServerToDevice()
+        {
+            string interfaceName = astarteMockDevice.InterfaceServerData;
+            Console.WriteLine("Test for individual datastreams"
+            + "in the direction from server to device.");
```
Comment (nedimtokic) [inline AstarteDeviceSDKCSharpE2E.Tests/AstarteDatastreamDeviceTests.cs:160]:
```suggestion
            + " in the direction from server to device.");
```

File: AstarteDeviceSDKCSharpE2E.Tests/AstartePropertyDeviceTests.cs [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+
+        /// <summary>
+        /// Test for individual properties in the direction from server to device.
+        /// </summary>
+        /// <returns></returns>
+        [Fact, TestPriority(4)]
+        public async Task PropertiesFromServerToDevice()
+        {
+            Console.WriteLine("Test for individual properties in"
+            + "the direction from server to device.");
```
Comment (nedimtokic) [inline AstarteDeviceSDKCSharpE2E.Tests/AstartePropertyDeviceTests.cs:114]:
```suggestion
            + " the direction from server to device.");
```

**Review by nedimtokic — CHANGES_REQUESTED** [review]

File: AstarteDeviceSDKCSharpE2E.Tests/AstarteAggregateDeviceTest.cs (hunk 8 above)
Comment (osmanhadzic) [reply AstarteDeviceSDKCSharpE2E.Tests/AstarteAggregateDeviceTest.cs]: I made changes.

File: AstarteDeviceSDKCSharpE2E.Tests/AstarteDatastreamDeviceTests.cs (hunk 9 above)
Comment (osmanhadzic) [reply AstarteDeviceSDKCSharpE2E.Tests/AstarteDatastreamDeviceTests.cs]: I made changes.

File: AstarteDeviceSDKCSharpE2E.Tests/AstarteDatastreamDeviceTests.cs (hunk 10 above)
Comment (osmanhadzic) [reply AstarteDeviceSDKCSharpE2E.Tests/AstarteDatastreamDeviceTests.cs]: I made changes.

File: AstarteDeviceSDKCSharpE2E.Tests/AstartePropertyDeviceTests.cs (hunk 11 above)
Comment (osmanhadzic) [reply AstarteDeviceSDKCSharpE2E.Tests/AstartePropertyDeviceTests.cs]: I made changes.

**Review by nedimtokic — APPROVED** [review]

File: AstarteDeviceSDKCSharpE2E.Tests/AstarteMockData/AstarteMockDevice.cs (hunk 5 above)
Comment (harlem88) [reply AstarteDeviceSDKCSharpE2E.Tests/AstarteMockData/AstarteMockDevice.cs:257]: @osmanhadzic Add these changes.

**Review by harlem88 — CHANGES_REQUESTED** [review]

**Review by harlem88 — APPROVED** [review]

**Pull request merged**